src/monmove.cpp

```diff
@@ -1381,6 +1527,73 @@ bool monster::attack_at( const tripoint &p )
     return false;
 }
 
+bool monster::pickup_at( const tripoint &p, item_location &target )
+{
+    //If called through normal pathfinding, then clear looting effect
+    if( p == goal ) {
+        remove_effect( effect_looting );
+
+    }
+
+    //Item has moved...
+    if( p != target.position() ) {
+        return false;
+    }
+
+    inv.push_back( *target.get_item() );
+    item *stored_item = &inv.back();
+
+    //Get weight of the stack or item
+    units::mass weight = target->weight();
+    units::mass capacity = this->weight_capacity();
+
+    int amount_taken = 1;
+
+    int charges = target->charges;
+
+    if( charges >= 1 ) {
+        //Adjust volume and weight for units in a stack
+        units::mass weigh_each = weight / charges;
+        amount_taken = charges + ( weigh_each / capacity );
+
+        //Caps the amount taken to avoid taking large stacks.
+        //TODO: better way to cap the amount taken?
+        if( amount_taken > 2 ) {
+            amount_taken = 2;
+        }
+
+        target->mod_charges( -amount_taken );
+        inv.back().charges = amount_taken;
+
+        //If we've taken all the stack, let's remove the item
+        if( amount_taken == charges ) {
+            target.remove_item();
+        }
+    } else {
+        //Remove item from ground
+        target.remove_item();
+    }
+
+    //Successfully taken any?
+    if( amount_taken >= 1 ) {
+        //Notify the player
+        if( g->u.sees( *this ) ) {
+            add_msg( m_warning, "%1$s grabs %2$s!", name().c_str(), stored_item->display_name() );
+        }
+
+
+        mod_moves( -100 );
+        return true;
+    } else {
+        //Failed to take any items, so remove the last item added to inventory.
+        inv.pop_back();
+    }
```

Comment:
The if-else structure is a bit confusing:
```C++
if( something ) {
    foo ...
    return;
} else {
    bar ...
}
stuff ...
return;
```
This looks as if there are three code segments (foo, bar, stuff), and either foo or bar is run (depending on the condition) and stuff is run always afterwards. But that's not the case (no return in the else-branch).

Please consider structuring it like this:
```C++
if( something ) {
    foo ...
    return;
}
bar ...
stuff ...
return;
```
Or this:
```C++
if( something ) {
    foo ...
    return;
} else {
    bar ...
    stuff ...
    return;
}
```